src/ck-ldb/TreeBuilder.h

```diff
@@ -0,0 +1,369 @@
+#ifndef TREEBUILDER_H
+#define TREEBUILDER_H
+
+#include "TreeLB.h"
+#include "TreeLevel.h"
+#include "spanningTree.h"
+#include <algorithm>
+
+extern int quietModeRequested;
+
+class LBTreeBuilderCommon : public LBTreeBuilder
+{
+ public:
+  virtual ~LBTreeBuilderCommon() {}
+
+ protected:
+  void reset(uint8_t num_levels, std::vector<LevelLogic*>& logic,
+             std::vector<int>& comm_parent, std::vector<std::vector<int>>& comm_children,
+             std::vector<LevelLogic*>& comm_logic)
+  {
+    logic.resize(num_levels);
+    comm_parent.resize(num_levels);
+    comm_children.resize(num_levels);
+    comm_logic.resize(num_levels);
+
+    std::fill(logic.begin(), logic.end(), nullptr);
+    std::fill(comm_parent.begin(), comm_parent.end(), -1);
+    for (auto& children : comm_children) children.clear();
+    std::fill(comm_logic.begin(), comm_logic.end(), nullptr);
+  }
+};
+
+class PE_Root_Tree : public LBTreeBuilderCommon
+{
+ public:
+  virtual ~PE_Root_Tree() {}
+
+  virtual uint8_t build(std::vector<LevelLogic*>& logic, std::vector<int>& comm_parent,
+                        std::vector<std::vector<int>>& comm_children,
+                        std::vector<LevelLogic*>& comm_logic, json& config)
+  {
+    uint8_t L = 2;  // num levels
+    reset(L, logic, comm_parent, comm_children, comm_logic);
+
+    int rootPE = 0;
+    try
```

Comment:
Can we avoid using exceptions? We have generally avoided using them entirely in the runtime